Resolution check

nike4613 opened the pull request:
Alright. This time the storing is done in `getSelectedQuality()`, and now it has permissions to `*.vrv.co`.
Priority of qualities is this: specified in URL > stored > selected on webpage.
`getSelectedQuality()` is now `async` by necessity of using the storage API, as well as `_run()`, however it does not seem to be noticable.

I will try to keep this branch updated with needed URL permissions as I find them.

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,65 @@ export function getAutoPlay(url: string): boolean|undefined {
 
-export function getSelectedQuality(): string|undefined {
+let storeQuality: string|undefined = ""; // empty string means this hasn't run yet
+export async function updateQualitySettings(): void {
```
Causes error due to an async method requiring the return type to be a Promise.

```TypeScript
export async function updateQualitySettings(): Promise<void> {
```

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,65 @@ export function getAutoPlay(url: string): boolean|undefined {
+export async function updateQualitySettings(): void {
+  let quality: string|undefined; 
+  
+  /**/
```
Empty comment?

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,65 @@ export function getAutoPlay(url: string): boolean|undefined {
+  const storage = container.get<IStorage>(IStorageSymbol);
+  // get and check saved quality
+  let savedQuality: string|undefined = await storage.get<string>("resolution");
+  if (savedQuality === undefined || (qualityOverride !== undefined && qualityOverride != savedQuality)) {
```
always use triple equal `qualityOverride != savedQuality` except in very specific cases.

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,65 @@ export function getAutoPlay(url: string): boolean|undefined {
-    return token.substring(10);
+    quality = token.match(/^showmedia\.(\d{2,3}0p)$/)[1]; // regex match, first capture (2-3 digits followed by 0p)
+  }
+  quality = undefined;
```
Don't set quality to undefined as the above work will go to waste otherwise. Also it will be used further down.

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,65 @@ export function getAutoPlay(url: string): boolean|undefined {
+  if (qualityOverride !== undefined)
+    quality = qualityOverride;
+  
+  //setTimeout(() => {
```
Remove the commented code

YePpHa commented on src/styles/player/index.scss:
```diff
@@ -56,4 +56,30 @@
+  }
+}
+
+#html5_loading_text {
```
rename to `.html5-loading-text`

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,65 @@ export function getAutoPlay(url: string): boolean|undefined {
+      targetEl.classList.replace('default-button','dark-button');
+      targetEl.classList.add('selected');
+    }
+  //}, 50); // postpone updating of buttons for a few milliseconds to return earlier
```
Remove the commented code

YePpHa commented on src/app/bootstrap.tsx:
```diff
@@ -11,16 +11,48 @@ import { bindCrossHttpClientAsDefault } from './config';
+  
+  if (stage > hightestRunState) {
+    for (let i = hightestRunState+1; i <= stage; i++) {
+      runState(i);
```
add await: `await runState(i);`

YePpHa commented on src/app/bootstrap.tsx:
```diff
@@ -11,16 +11,48 @@ import { bindCrossHttpClientAsDefault } from './config';
+  }
+  
+  if (stage > hightestRunState) {
+    for (let i = hightestRunState+1; i <= stage; i++) {
```
You can use the example at https://stackoverflow.com/a/39372911 to loop through the enum that I suggested you should use.

YePpHa commented on src/app/bootstrap.tsx:
```diff
@@ -11,16 +11,48 @@ import { bindCrossHttpClientAsDefault } from './config';
+  readyState();
+}
+
+var bootstrapper: Bootstrap;
```
nit: don't use var use let

YePpHa commented on src/app/bootstrap.tsx:
```diff
@@ -11,16 +11,48 @@ import { bindCrossHttpClientAsDefault } from './config';
+
+var bootstrapper: Bootstrap;
+
+var hightestRunState: number = 0;
```
nit: don't use var use let

YePpHa commented on src/app/bootstrap.tsx:
```diff
@@ -11,16 +11,48 @@ import { bindCrossHttpClientAsDefault } from './config';
-    const fn = function() {
-      if (document.readyState === "complete") {
-        document.removeEventListener("readystatechange", fn, false);
+  // replaced old checks with a system that runs something at 'loading', 'interactive', and 'complete'
```
nit: comment not needed as the one reading the code wouldn't know what came before.

nike4613 commented:
Alrighty, thanks for the suggestions. I just went and implemented all of them. I am still using numbers with the enum because I don't trust the iteration order otherwise.

YePpHa commented:
👍 Everything seems fine. I will do some testing to verify everything is in working order.

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,64 @@ export function getAutoPlay(url: string): boolean|undefined {
+  let cselEl = document.querySelector("a.selected[token^=showmedia\\.]")!;
+  let targetEl = document.querySelector("a[token^=showmedia\\." + quality + "]")!;
+  
+  if (cselEl != targetEl) {
```
use `!==` to be consistent in how we compare two values

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -35,16 +37,64 @@ export function getAutoPlay(url: string): boolean|undefined {
+    // tslist doesn't seem to know about replace
+    // it is supported in everything but Opera, Safari and IE
+    // see: https://developer.mozilla.org/en-US/docs/Web/API/Element/classList
+    cselEl.classList.replace('dark-button','default-button');
```
I looked into this a bit and it might be better to not use `replace` as some users might use a version of Firefox older than v49. However, I did look into using [core-js](https://github.com/zloirock/core-js), but as far as I could see they don't polyfill the `replace` method. It's probably better to just write our own replace method and use that one. The following code almost keeps to the specs of replace, however, it doesn't keep the ordering of the replaced class.

```TypeScript
function replace(list, oldClass, newClass): void {
  if (!list.contains(oldClass)) return;

  list.remove(oldClass);
  list.add(newClass);
}

replace(cselEl.classList, 'dark-button', 'default-button');
cselEl.classList.remove('selected');
replace(targetEl.classList, 'default-button', 'dark-button');
targetEl.classList.add('selected');
```

Also with TypeScript not knowing `replace` is because the declaration file for `DOMTokenList` doesn't contain the method. You can actually add it by adding the following to `StandardPlayer.ts`:
```TypeScript
declare global {
  interface DOMTokenList {
    replace(oldClass: string, newClass: string): void;
  }
}
```

nike4613 commented:
I just added a `replace` polyfill and updated some permissions to allow all VRV subdomains.
I also performed an incredibly messy rebase so that this branch would be on version 0.9.4.

YePpHa commented on src/app/libs/polyfill/global.ts:
```diff
@@ -0,0 +1,27 @@
+  function replace(this: DOMTokenList, oldClass: string, newClass: string): void {
+    let list = new Array(this.length);
+    for (let el in this) {
+      if (el == oldClass)
```
use triple equals (`===`)

YePpHa commented on src/app/libs/polyfill/global.ts:
```diff
@@ -0,0 +1,27 @@
+DOMTokenList.prototype.replace = DOMTokenList.prototype.replace || 
+  // replace preserving order
+  function replace(this: DOMTokenList, oldClass: string, newClass: string): void {
+    let list = new Array(this.length);
```
Most of this can be reduced to the following code as the DOMTokenList is actually array-like which means the [Array.from](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from) works:
```TypeScript
const list = Array.from(this).map(x => x === oldClass ? newClass : x);
this.empty();
this.add(...list);
```

YePpHa commented on src/app/libs/polyfill/global.ts:
```diff
@@ -0,0 +1,27 @@
+  }
+}
+
+export default undefined;
```
don't export undefined as it's not needed.

nike4613 replied:
Unfortunately,  TypeScript doesn't think this is a module without this, and it doesn't let me define any types in the global scope outside of a module.

YePpHa commented on src/app/player/StandardPlayer.ts:
```diff
@@ -1,4 +1,7 @@
 import * as parseUrl from 'url-parse';
+import { IStorageSymbol, IStorage } from '../storage/IStorage';
+import container from "../../config/inversify.config";
+import polyfill from "../libs/polyfill/global";
```
just do `import "../libs/polyfill/global";`

YePpHa replied:
rename it from `global` to `DOMTokenList`

YePpHa commented:
I just got a question why is it like you completely removed everything and committed it again? This makes it kind of hard to find the changes you've made since my last review.

nike4613 commented:
Thanks again for the feedback. 

The reason that everything was re-committed was that I made a major mistake when trying to merge in recent changes to this branch, and ended up having to rebase to master. Sorry about that. It's more than a bit of a mess.

nike4613 commented:
Thanks again for the feedback.

The reason that everything was re-committed was that I made a major mistake when trying to merge in recent changes to this branch, and ended up having to rebase to master. Sorry about that. It's more than a bit of a mess.

YePpHa commented:
@nike4613 it's possible to revert commits if you screw something up.
